### PR TITLE
feat(types): defining models, offers stricter type safety by default

### PR DESCRIPTION
**Pull request by Keimeno**
### Pull Request check-list

- [X] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Have you added new tests to prevent regressions?
- [X] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [X] Did you update the typescript typings accordingly (if applicable)?
- [X] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description of change

***This PR currently introduces a breaking change***

#### Better API design for strictly defining models

The best way to offer type safety for a Sequelize model that is defined with the `define` method is to create a `ModelAttributes` and a `CreationAttributes` interface like the following:

```ts
interface InstrumentAttributes {
	id: number;
	type: string;
	purchaseDate: Date;
}

interface InstrumentCreationAttributes
	extends Optional<InstrumentAttributes, "id"> {}

const Instrument: ModelDefined<
  InstrumentAttributes, 
  InstrumentCreationAttributes
> = sequelize.define("instrument", {
	type: DataTypes.STRING,
	purchaseDate: DataTypes.DATE,
});
```

This generally works, but it has one flaw. We must always import the `ModelDefined` type. A better solution for this is to instead write:

```ts
const Instrument = sequelize.define<InstrumentAttributes, InstrumentCreationAttributes>("instrument", {
	type: DataTypes.STRING,
	purchaseDate: DataTypes.DATE,
});
```

This PR introduces this new API design.

#### Stricter type safety if not strictly passing `ModelAttributes` & `CreationAttributes`

Currently, when defining a model, without passing the `ModelAttributes` and `CreationAttributes`, we run into the following Problem; The return type is: `ModelCtor<Model>`, which does not offer type safety for attributes at all, meaning if we write this: `Instrument.create({ type: 'foo' })`, it won't give any type suggestions for the attribute `type`.

With this PR and out above example, but with the generics `InstrumentAttributes` and `InstrumentCreationAttributes` removed, `sequelize.define` would return the following type: 

```ts
ModelDefined<
    | Record<string, unknown>
    | ({
        type: unknown; 
        purchaseDate: unknown;
      } & Record<string, unknown>),
    {
      type: unknown; 
      purchaseDate: unknown;
    } & Record<string, unknown>
  >
```

With this addition, it is known that the attributes `type` and `purchaseDate` exist, but are of type `unknown`. It is also possible to add `attributes`, that were not explicitly specified during the model definition, such as `id`.

#### Summing up

This PR offers a more accessible API design, at the cost of introducing a breaking change. The breaking change is, that you can't strictly define the `ModelAttributes` and `CreationAttributes`, as you could before. I'm currently investigating this, to see if there's any way we can offer this backwards compatibility, and I'm open to ideas.

### CODE REVIEW

**Comment by Keimeno** [comment]
There may be more incompatibilities introduced by this than I initially thought. Closing this now, but this might get relevant for v7
